crates/solidity/outputs/cargo/crate/src/backend/types.rs

```diff
@@ -335,32 +723,46 @@ impl Type {
         }
     }
 
-    #[must_use]
-    pub fn with_data_location(&self, data_location: Option<DataLocation>) -> Self {
+    pub fn is_inherited_location(&self) -> bool {
+        self.data_location()
+            .is_some_and(|location| location == DataLocation::Inherited)
+    }
+
+    pub fn can_return_from_getter(&self) -> bool {
         match self {
-            Self::Array {
-                element_type,
-                location,
-            } => Self::Array {
-                element_type: *element_type,
-                location: data_location.unwrap_or(*location),
-            },
-            Self::Bytes { location } => Self::Bytes {
-                location: data_location.unwrap_or(*location),
-            },
-            Self::String { location } => Self::String {
-                location: data_location.unwrap_or(*location),
-            },
-            Self::Struct {
-                definition_id,
-                location,
-            } => Self::Struct {
-                definition_id: *definition_id,
-                location: data_location.unwrap_or(*location),
-            },
-            _ => self.clone(),
+            Type::Address { .. }
+            | Type::Boolean
+            | Type::ByteArray { .. }
+            | Type::Bytes { .. }
+            | Type::Contract { .. }
+            | Type::Enum { .. }
+            | Type::FixedPointNumber { .. }
+            | Type::Integer { .. }
+            | Type::Interface { .. }
+            | Type::String { .. }
+            | Type::UserDefinedValue { .. } => true,
+
+            Type::Array { .. }
+            | Type::Function(_)
+            | Type::Mapping { .. }
+            | Type::Literal(_)
+            | Type::Struct { .. }
+            | Type::Tuple { .. }
+            | Type::Void => false,
         }
     }
+
+    pub fn is_literal_number(&self) -> bool {
+        matches!(
+            self,
+            Type::Literal(
+                LiteralKind::Zero
+                    | LiteralKind::DecimalInteger
+                    | LiteralKind::HexInteger { .. }
+                    | LiteralKind::Rational
+            )
+        )
+    }
 }
 
```

Comment:
same suggestion for this module: splitting different types into sub-modules for isolation/readability.